File: src/main/java/no/fintlabs/mapping/EgrunnervervSakInstanceMappingService.java
package no.fintlabs.mapping;

import lombok.extern.slf4j.Slf4j;
import no.fintlabs.ResourceRepository;
import no.fintlabs.exceptions.ArchiveResourceNotFoundException;
import no.fintlabs.exceptions.NonMatchingEmailDomainWithOrgIdException;
import no.fintlabs.gateway.instance.InstanceMapper;
import no.fintlabs.gateway.instance.model.instance.InstanceObject;
import no.fintlabs.models.EgrunnervervSakInstance;
import no.fintlabs.models.EgrunnervervSakKlassering;
import no.fintlabs.models.EgrunnervervSaksPart;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.HashMap;
import java.util.Map;

@Service
@Slf4j
public class EgrunnervervSakInstanceMappingService implements InstanceMapper<EgrunnervervSakInstance> {

    @Value("${fint.flyt.egrunnerverv.checkSaksansvarligEpost:true}")
    boolean checkSaksansvarligEpost;

    @Value("${fint.flyt.egrunnerverv.checkEmailDomain:true}")
    boolean checkEmailDomain;
    private final ResourceRepository resourceRepository;

    @Value("${fint.org-id}")
    private String orgId;

    private final FormattingUtilsService formattingUtilsService;

    public EgrunnervervSakInstanceMappingService(
            ResourceRepository resourceRepository,
            FormattingUtilsService formattingUtilsService
    ) {
        this.resourceRepository = resourceRepository;
        this.formattingUtilsService = formattingUtilsService;
    }

    @Override
    public Mono<InstanceObject> map(Long sourceApplicationId, EgrunnervervSakInstance egrunnervervSakInstance) {
        return Mono.defer(() -> {

            String saksansvarligEpostTrimmed = egrunnervervSakInstance.getSaksansvarligEpost().trim();

            if (checkEmailDomain) {
                String domain = formattingUtilsService.extractEmailDomain(saksansvarligEpostTrimmed);
                if (!domain.equals(orgId)) {
                    throw new NonMatchingEmailDomainWithOrgIdException(domain, orgId);
                }
            }

            String saksansvarlig = "";

            if (checkSaksansvarligEpost) {
                saksansvarlig = resourceRepository
                        .getArkivressursHrefFromPersonEmail(saksansvarligEpostTrimmed)
                        .orElseThrow(() -> new ArchiveResourceNotFoundException(
                                        saksansvarligEpostTrimmed
                                )
                        );
            }

            Map<String, String> valuePerKey = new HashMap<>();
            valuePerKey.put("sys_id", egrunnervervSakInstance.getSysId());
            valuePerKey.put("knr", egrunnervervSakInstance.getKnr());
            valuePerKey.put("gnr", egrunnervervSakInstance.getGnr());
            valuePerKey.put("bnr", egrunnervervSakInstance.getBnr());
            valuePerKey.put("fnr", egrunnervervSakInstance.getFnr());
            valuePerKey.put("snr", egrunnervervSakInstance.getSnr());
            valuePerKey.put("takstnummer", egrunnervervSakInstance.getTakstnummer());
            valuePerKey.put("tittel", egrunnervervSakInstance.getTittel());
            valuePerKey.put("saksansvarligEpost", saksansvarligEpostTrimmed);
            valuePerKey.put("saksansvarlig", saksansvarlig);
            valuePerKey.put("eierforholdsnavn", egrunnervervSakInstance.getEierforholdsnavn());
            valuePerKey.put("eierforholdskode", egrunnervervSakInstance.getEierforholdskode());
            valuePerKey.put("prosjektnr", egrunnervervSakInstance.getProsjektnr());
            valuePerKey.put("prosjektnavn", egrunnervervSakInstance.getProsjektnavn());
            valuePerKey.put("kommunenavn", formattingUtilsService.formatKommunenavn(egrunnervervSakInstance.getKommunenavn()));
            valuePerKey.put("adresse", egrunnervervSakInstance.getAdresse());

            return Mono.just(
                    InstanceObject.builder()
                            .valuePerKey(valuePerKey)
                            .objectCollectionPerKey(Map.of(
                                    "saksparter", egrunnervervSakInstance.getSaksparter()
                                            .stream()
                                            .map(this::toInstanceObject)
                                            .toList(),
                                    "klasseringer", egrunnervervSakInstance.getKlasseringer()
                                            .stream()
                                            .map(this::toInstanceObject)
                                            .toList()
                            ))
                            .build()
            );
        });
    }

    private InstanceObject toInstanceObject(EgrunnervervSaksPart egrunnervervSaksPart) {
        return InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "navn", egrunnervervSaksPart.getNavn(),
                        "organisasjonsnummer", egrunnervervSaksPart.getOrganisasjonsnummer(),
                        "epost", egrunnervervSaksPart.getEpost(),
                        "telefon", egrunnervervSaksPart.getTelefon(),
                        "postadresse", egrunnervervSaksPart.getPostadresse(),
                        "postnummer", egrunnervervSaksPart.getPostnummer(),
                        "poststed", egrunnervervSaksPart.getPoststed()
                ))
                .build();
    }

    private InstanceObject toInstanceObject(EgrunnervervSakKlassering egrunnervervSakKlassering) {
        return InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "ordningsprinsipp", egrunnervervSakKlassering.getOrdningsprinsipp(),
                        "ordningsverdi", egrunnervervSakKlassering.getOrdningsverdi(),
                        "beskrivelse", egrunnervervSakKlassering.getBeskrivelse(),
                        "sortering", egrunnervervSakKlassering.getSortering(),
                        "untattOffentlighet", egrunnervervSakKlassering.getUntattOffentlighet()
                ))
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/instance/mapping/EgrunnervervSakInstanceMappingService.kt
package no.novari.flyt.egrunnerverv.gateway.instance.mapping

import no.novari.flyt.egrunnerverv.gateway.exception.ArchiveResourceNotFoundException
import no.novari.flyt.egrunnerverv.gateway.exception.NonMatchingEmailDomainWithOrgIdException
import no.novari.flyt.egrunnerverv.gateway.infrastructure.slack.SlackAlertService
import no.novari.flyt.egrunnerverv.gateway.instance.ResourceRepository
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakInstance
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSakKlassering
import no.novari.flyt.egrunnerverv.gateway.instance.model.EgrunnervervSaksPart
import no.novari.flyt.gateway.instance.InstanceMapper
import no.novari.flyt.gateway.instance.model.File
import no.novari.flyt.gateway.instance.model.instance.InstanceObject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class EgrunnervervSakInstanceMappingService(
    private val resourceRepository: ResourceRepository,
    private val formattingUtilsService: FormattingUtilsService,
    private val slackAlertService: SlackAlertService,
    @param:Value("\${novari.flyt.egrunnerverv.checkSaksansvarligEpost:true}")
    private val checkSaksansvarligEpost: Boolean,
    @param:Value("\${novari.flyt.egrunnerverv.checkEmailDomain:true}")
    private val checkEmailDomain: Boolean,
    @param:Value("\${fint.org-id}")
    private val orgId: String,
) : InstanceMapper<EgrunnervervSakInstance> {
    override fun map(
        sourceApplicationId: Long,
        incomingInstance: EgrunnervervSakInstance,
        persistFile: (File) -> UUID,
    ): InstanceObject {
        val saksansvarligEpostFormatted = formattingUtilsService.formatEmail(incomingInstance.saksansvarligEpost)

        if (checkEmailDomain) {
            val domain =
                formattingUtilsService.extractEmailDomain(saksansvarligEpostFormatted)
                    ?: throw IllegalArgumentException("Invalid email address")
            if (domain != orgId) {
                throw NonMatchingEmailDomainWithOrgIdException(domain, orgId)
            }
        }

        val saksansvarlig =
            if (checkSaksansvarligEpost) {
                resourceRepository.getArkivressursHrefFromPersonEmail(saksansvarligEpostFormatted)
                    ?: throw ArchiveResourceNotFoundException(saksansvarligEpostFormatted, slackAlertService)
            } else {
                ""
            }

        return InstanceObject(
            valuePerKey =
                linkedMapOf(
                    "sys_id" to incomingInstance.sysId,
                    "knr" to incomingInstance.knr,
                    "gnr" to incomingInstance.gnr,
                    "bnr" to incomingInstance.bnr,
                    "fnr" to incomingInstance.fnr,
                    "snr" to incomingInstance.snr,
                    "takstnummer" to incomingInstance.takstnummer,
                    "tittel" to incomingInstance.tittel,
                    "saksansvarligEpost" to saksansvarligEpostFormatted,
                    "saksansvarlig" to saksansvarlig,
                    "eierforholdsnavn" to incomingInstance.eierforholdsnavn,
                    "eierforholdskode" to incomingInstance.eierforholdskode,
                    "prosjektnr" to incomingInstance.prosjektnr,
                    "prosjektnavn" to incomingInstance.prosjektnavn,
                    "kommunenavn" to (formattingUtilsService.formatKommunenavn(incomingInstance.kommunenavn) ?: ""),
                    "adresse" to incomingInstance.adresse,
                ),
            objectCollectionPerKey =
                mutableMapOf(
                    "saksparter" to incomingInstance.saksparter.map(::toInstanceObject),
                    "klasseringer" to incomingInstance.klasseringer.map(::toInstanceObject),
                ),
        )
    }

    private fun toInstanceObject(saksPart: EgrunnervervSaksPart): InstanceObject =
        InstanceObject(
            valuePerKey =
                mapOf(
                    "navn" to saksPart.navn,
                    "organisasjonsnummer" to saksPart.organisasjonsnummer,
                    "epost" to saksPart.epost,
                    "telefon" to saksPart.telefon,
                    "postadresse" to saksPart.postadresse,
                    "postnummer" to saksPart.postnummer,
                    "poststed" to saksPart.poststed,
                ),
        )

    private fun toInstanceObject(klassering: EgrunnervervSakKlassering): InstanceObject =
        InstanceObject(
            valuePerKey =
                mapOf(
                    "ordningsprinsipp" to klassering.ordningsprinsipp,
                    "ordningsverdi" to klassering.ordningsverdi,
                    "beskrivelse" to klassering.beskrivelse,
                    "sortering" to klassering.sortering,
                    "untattOffentlighet" to klassering.untattOffentlighet,
                ),
        )
}
